### PR TITLE
[pf serving] add fastapi engine support

### PR DESCRIPTION
**Pull request by wxpjimmy**
# Description

- refactor the existing flask serving app to make sure common code is engine agnostic
- add new fastapi serving engine and related tests

refactored code structure introduction:   
`/promptflow/core/_serving`: 
 - `app.py`:  serving app entry,  can control which serving app to use with "engine=xxx (fastapi/flask" param
 - `app_base.py`:  common app initialization logic in base app  
 - `extension/`:   serving engine extension module, which can be used to customize platform specific serving logic.
 - `monitor/` :  serving metrics & trace monitor
 - `resources/`:  static resources used internally 
 - `static/`:  static web resources
 - `v1/`:  flask engine related code
 - `v2/`:  fastapi engine related code

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow-core/promptflow/core/_serving/v2/routers/score.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        # TODO: refine this once we can directly set the input/output log level to DEBUG in flow_invoker.
+        disable_data_logging = logger.level >= logging.INFO
+        span = trace.get_current_span()
+        if span == INVALID_SPAN:
+            # no parent span, try to extract trace context from request
+            logger.info("No parent span found, try to extract trace context from request.")
+            ctx = try_extract_trace_context(logger, request.headers)
+        else:
+            ctx = None
+        # TODO: change to use async once ScriptExecutor support async run
```
Comment (thy09) [inline src/promptflow-core/promptflow/core/_serving/v2/routers/score.py:53]: I think the main branch has already supported async run?

Comment (wxpjimmy) [reply src/promptflow-core/promptflow/core/_serving/v2/routers/score.py:53]: Let me check

Comment (wxpjimmy) [reply src/promptflow-core/promptflow/core/_serving/v2/routers/score.py]: confirmed and changed to use async run for flex flow

File: src/promptflow-core/pyproject.toml [hunk 2]
```diff
@@ -49,8 +49,8 @@ jsonschema = ">=4.0.0,<5.0.0"  # used to validate tool
 filetype = ">=1.2.0"  # used to detect the mime type for multi-media input
 flask = ">=2.2.3,<4.0.0"  # Serving endpoint requirements
 python-dateutil = ">=2.1.0,<3.0.0"  # Contracts RunInfo
-# ========executor-service dependencies========
-fastapi = ">=0.109.0,<1.0.0" # used to build web executor server
+# ========fastapi server dependencies========
+fastapi = ">=0.109.0,<1.0.0" # used to build fastapi server
```
Comment (wangchao1230) [inline src/promptflow-core/pyproject.toml:53]: shall this keep in extra?

File: src/promptflow-core/promptflow/core/_serving/app.py [hunk 3]
```diff
@@ -2,281 +2,48 @@
...
-        span = trace.get_current_span()
-        if span == INVALID_SPAN:
-            # no parent span, try to extract trace context from request
-            ctx = try_extract_trace_context(logger)
-        else:
-            ctx = None
-        token = context.attach(ctx) if ctx else None
+def create_app(**kwargs):
+    engine = kwargs.pop("engine", "flask")
+    if engine == "flask":
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/core/_serving/app.py:12]: User current not setting this engine. Do we expect to switch to this by default? Or user does that in their Dockerfile?

Shall we add doc for this new engine and when to use this?

File: src/promptflow-core/promptflow/core/_serving/flow_invoker.py [hunk 4]
```diff
@@ -252,7 +251,12 @@ async def invoke_async(self, data: dict, run_id=None, disable_input_output_loggi
             data, run_id=run_id, disable_input_output_logging=disable_input_output_logging
         )
         # Get base64 for multi modal object
-        resolved_outputs = self._convert_multimedia_data_to_base64(result)
+        output_dict = convert_eager_flow_output_to_dict(result.output)
```
Comment (brynn-code) [inline src/promptflow-core/promptflow/core/_serving/flow_invoker.py:254]: @D-W- Please double confirm here

Comment (D-W-) [reply src/promptflow-core/promptflow/core/_serving/flow_invoker.py:254]: Original code was added in sync invoker. seems this one was added to async one

File: src/promptflow-core/pyproject.toml (hunk 2 above)
Comment (crazygao) [reply src/promptflow-core/pyproject.toml:53]: Also, promptflow-devkit project also need this as optional includes.

Comment (wxpjimmy) [reply src/promptflow-core/pyproject.toml:53]: it's included in core already, no need to put in extra any more

File: src/promptflow-core/promptflow/core/_serving/app.py (hunk 3 above)
Comment (wxpjimmy) [reply src/promptflow-core/promptflow/core/_serving/app.py:12]: the default value currently is still flask,  for short term we will provide environment variable for customer to choose switching to use fastapi,  once we have enough confidence we plan to switch to use fastapi as default.

As for doc, definitely we will update the doc for  how to enable this.

File: src/promptflow-core/promptflow/core/_serving/flow_invoker.py (hunk 4 above)
Comment (wxpjimmy) [reply src/promptflow-core/promptflow/core/_serving/flow_invoker.py:254]: yes, for fastapi we will all change to use this async api, so this is to make sure the logic is aligned with the sync call

File: src/promptflow-core/pyproject.toml (hunk 2 above)
Comment (wxpjimmy) [reply src/promptflow-core/pyproject.toml:53]: we keep the extra "executor-service", just leave it empty.

**Review by brynn-code — APPROVED** [review]

File: src/promptflow-core/pyproject.toml (hunk 2 above)
Comment (wangchao1230) [reply src/promptflow-core/pyproject.toml:53]: We can keep this for now. But need to resolve in future.

**Review by wangchao1230 — APPROVED** [review]

File: src/promptflow-core/pyproject.toml (hunk 2 above)
Comment (wxpjimmy) [reply src/promptflow-core/pyproject.toml:53]: make sense to me, let's have a separate discussion about how to better manage serving & core dependencies.

**Review by brynn-code — APPROVED** [review]

File: src/promptflow-devkit/promptflow/_internal/__init__.py [hunk 5]
```diff
@@ -107,11 +107,10 @@
 from promptflow.core._serving.utils import (
     get_output_fields_to_remove,
     get_sample_json,
-    handle_error_to_response,
     load_request_data,
-    streaming_response_required,
     validate_request_data,
 )
+from promptflow.core._serving.v1.utils import handle_error_to_response, streaming_response_required
```
Comment (thy09) [inline src/promptflow-devkit/promptflow/_internal/__init__.py:113]: We may remove such references since runtime no longer use this.

**Review by thy09 — APPROVED** [review]

File: src/promptflow-devkit/promptflow/_internal/__init__.py (hunk 5 above)
Comment (wxpjimmy) [reply src/promptflow-devkit/promptflow/_internal/__init__.py:113]: sure, will double confirm with Brynn and remove this if they are not needed anymore.

**Pull request merged**